src/pyfmi/fmi_algorithm_drivers.py

```diff
@@ -543,6 +552,73 @@ def _set_solver_options(self):
+            except:
+                support_elapsed_time=False
+
+            support_event_indicators = (solver_name=="CVode" or solver_name=="Radau5ODE" or solver_name=="LSODAR" or solver_name=="ImplicitEuler")
+
+            states_list = self.model.get_states_list() if isinstance(self.model, fmi.FMUModelME2) else []
+            
+            if solver_name != "ExplicitEuler":
+                try:
+                    rtol = self.solver_options['rtol']
+                    atol = self.solver_options['atol']
+                except KeyError:
+                    rtol, atol = self.model.get_tolerances()
+                self._diagnostics_param_names.append((diagnostics_start_name+"relative_tolerance", "Relative solver tolerance."))
+                self._diagnostic_param_values.append(rtol)
+                
+                for idx, state in enumerate(states_list):
+                    self._diagnostics_param_names.append((diagnostics_start_name+"absolute_tolerance."+state, "Absolute tolerance for "+state+"."))
+                    self._diagnostic_param_values.append(atol[idx])
+
+            self._diagnostics_var.append((diagnostics_start_name+"step_time","Step time"))
+            self._diagnostics_var_start_values.append(self.start_time)
+            if support_elapsed_time:
+                self._diagnostics_var.append((diagnostics_start_name+"cpu_time", "Cpu time for current step."))
+                self._diagnostics_var_start_values.append(0.0)
+            if support_solver_order:
+                self._diagnostics_var.append((diagnostics_start_name+"solver_order", "Solver order for CVode"))
+                self._diagnostics_var_start_values.append(0.0)
+            if support_state_errors:
+                for state in states_list:
+                    self._diagnostics_var.append((diagnostics_start_name+"state_errors."+state, "State error for "+state+"."))
+                    self._diagnostics_var_start_values.append(0.0)
+            if support_event_indicators:
+                nof_states, nof_ei = self.model.get_ode_sizes()
+                ei_values = self.model.get_event_indicators() if nof_ei > 0 else []
+                for i in range(nof_ei):
+                    self._diagnostics_var.append((diagnostics_start_name+"event_info.indicator_"+str(i+1), "Value for event indicator {}.".format(i+1)))
+                    self._diagnostics_var_start_values.append(ei_values[i])
+                for i in range(nof_ei):
+                    self._diagnostics_var.append((diagnostics_start_name+"event_info.state_event_info.index_"+str(i+1), "Zero crossing indicator for event indicator {}".format(i+1)))
+                    self._diagnostics_var_start_values.append(0.0)
+                self._diagnostics_var.append((diagnostics_start_name+"event_info.event_type", "No event=-1, state event=0, time event=1")) 
+                self._diagnostics_var_start_values.append(-1) 
+

```

Comment:
Perhaps you should have the diagnostics values as a map name → value or similar instead of a list of names and a list of values? The current approach seems both more fragile and less logical. If you later need it as a list, it is easy enough to convert. I see that the “names” one is actually [(name, description)] - perhaps you could do {name: (description, value)} or [(name, description, value)]?
There are a lot of places where you pass on all three diagnostics_* arguments on to other functions. It gets very verbose and in the way of readability. You may want to consider an object holding all three of them - then you could just pass that around.
You could also combine both of these to have an object with a map (or whatever is used for diagnostics_param_names + diagnostic_param_values) and a list for diagnostics_var.

Reply:
Changed in commit below to use dicts instead.